File: 9.贪心算法/3.摆动序列/376.摆动序列.cpp
/*
 * @lc app=leetcode.cn id=376 lang=cpp
 *
 * [376] 摆动序列
 */
#include "tools.h"
// @lc code=start
class Solution {
public:
    int dp[1005][2];
    int wiggleMaxLength(vector<int>& nums) {
        memset(dp, 0, sizeof(dp));
        dp[0][0] = dp[0][1] = 1;
        for (int i = 1; i < nums.size(); i ++ )
        {
            dp[i][0] = dp[i][1] = 1;
            for (int j = 0; j < i; j ++ )
            {
                if (nums[j] > nums[i])
                {
                    dp[i][0] = max(dp[i][0], dp[j][1] + 1);
                }
            }
            for (int j = 0; j < i; j ++ )
            {
                if (nums[j] < nums[i])
                {
                    dp[i][1] = max(dp[i][1], dp[j][0] + 1);
                }
            }
        }
        return max(dp[nums.size() - 1][1], dp[nums.size() - 1][0]);
    }
};
// @lc code=end

class Solution {
    public:
        int wiggleMaxLength(vector<int>& nums) {
            int lastdiff = 0;
            int res = 1;
            for (int i = 1; i < nums.size(); i ++ )
            {
                int curdiff = nums[i] - nums[i - 1];
                if ((lastdiff >= 0 && curdiff < 0) || (lastdiff <= 0 && curdiff > 0))
                {
                    res ++;
                    lastdiff = curdiff;
                }
                else if (curdiff != 0) lastdiff = curdiff;
            }
            return res;
        }
    };

File: 9.贪心算法/3.摆动序列/376.摆动序列.py
#
# @lc app=leetcode.cn id=376 lang=python3
#
# [376] 摆动序列
#
from mytools import *
# @lc code=start
class Solution:
    def wiggleMaxLength(self, nums: List[int]) -> int:
        res = 1
        lastvis = 0
        for i in range(len(nums) - 1):
            curdiff = nums[i + 1] - nums[i]
            if (lastvis >= 0 and curdiff < 0) or (lastvis <= 0 and curdiff > 0):
                lastvis = curdiff
                res += 1
            elif curdiff != 0: lastvis = curdiff
        return res

# @lc code=end

